feeds/wlan-ap/opensync/src/src/sm/src/ubus_collector.c

```diff
@@ -541,6 +561,94 @@ static void ubus_collector_bss_cb(struct ubus_request *req, int type,
 	}
 }
 
+static void ubus_collector_chan_switch_events_cb(struct ubus_request *req, int type,
+				  struct blob_attr *msg)
+{
+
+	int error = 0;
+	int rem = 0;
+	struct blob_attr *tb_chan_event_lst[__CHANNEL_SWITCH_EVENT_MAX] = {};
+	struct blob_attr *tb_chan_event_table[__CHANNEL_SWITCH_MAX] = {};
+	struct blob_attr *tb_chan_event_tbl = NULL;
+	dpp_event_channel_switch_t *channel_event_record = NULL;
+
+	if (!msg)
+		return;
+
+	LOG(INFO, "ubus_collector: received ubus collector chan event message");
+
+	error = blobmsg_parse(ubus_collector_chan_switch_event_policy,
+			      __CHANNEL_SWITCH_EVENT_MAX, tb_chan_event_lst,
+			      blobmsg_data(msg), blobmsg_data_len(msg));
+	if (error || !tb_chan_event_lst[CHAN_SWITCH_EVENT]) {
+		LOG(INFO,
+		    "ubus_collector: failed while parsing chan event policy");
+		return;
+	}
+
+	/* itereate channel event list */
+	blobmsg_for_each_attr(tb_chan_event_tbl, tb_chan_event_lst[CHAN_SWITCH_EVENT], rem)
+	{
+		error = blobmsg_parse(channel_switch_event_policy,
+				      __CHANNEL_SWITCH_MAX, tb_chan_event_table,
+				      blobmsg_data(tb_chan_event_tbl),
+				      blobmsg_data_len(tb_chan_event_tbl));
+		if (error) {
+			LOG(INFO,
+			    "ubus_collector_ failed while parsing chan event policy");
+			continue;
+		}
+
+		channel_event_record = malloc(sizeof(dpp_event_channel_switch_t));
+
+		channel_event_record->channel_event.band = frequency_to_band(blobmsg_get_u32(tb_chan_event_table[CHANNEL_SWITCH_RADIO_NAME]));
+
+		channel_event_record->channel_event.reason = blobmsg_get_u8(tb_chan_event_table[CHANNEL_SWITCH_REASON]);
+		channel_event_record->channel_event.freq = blobmsg_get_u8(tb_chan_event_table[CHANNEL_SWITCH_FREQ]);
+		channel_event_record->channel_event.timestamp = blobmsg_get_u32(tb_chan_event_table[CHANNEL_SWITCH_TIMESPEC]);
+
+		ds_dlist_insert_tail(&g_event_report.channel_switch_list, channel_event_record);
+	}
+
+	return;
+
+
+}
+
+static void ubus_collector_complete_channel_switch_cb(struct ubus_request *req, int ret)
+{
+	LOG(DEBUG, "ubus_collector_complete_channel_switch_cb");
+	if (req)
+		free(req);
+}
+
+static void ubus_collector_hostapd_channel_switch_invoke(void *arg)
+{
+	uint32_t ubus_object_id = 0;
+	const char *object_path = "hostapd";
+	const char *hostapd_method = "get_chan_switch_events";
+	struct ubus_request *req = malloc(sizeof(struct ubus_request));
+
+	if (ubus_lookup_id(ubus, object_path, &ubus_object_id)) {
+		LOG(INFO, "ubus_collector: could not find ubus object %s",
+		    object_path);
+		free(req);
+		evsched_task_reschedule();
+		return;
+	}
+
+	LOG(INFO, "ubus_collector: requesting hostapd channel switch data");
+
+	ubus_invoke_async(ubus, ubus_object_id, hostapd_method, NULL, req);
+
+	req->data_cb = (ubus_data_handler_t)ubus_collector_chan_switch_events_cb;
+	req->complete_cb = (ubus_complete_handler_t)ubus_collector_complete_channel_switch_cb;
+
+	ubus_complete_request(ubus, req, 0);
+
+	evsched_task_reschedule();
+}
+
```

Comment:
This will not work. "get_chan_switch_events" is defined per bss. ubus object path should be hostapd.wlan0, hostapd.wlan1 and so on and not hostapd. This needs to be called for each bss.